front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts

```diff
@@ -125,280 +94,386 @@ const DEFAULT_TIME_LOCK: TimeLockDto = {
   timeFormatter: null,
 };
 
-/**
- * Build a search query to fetch all operational points from their UICs,
- * trigrams and IDs.
- */
-const buildOpQuery = (
-  infraId: number,
-  trainSchedules: TrainScheduleResult[]
-): SearchPayload | null => {
-  const pathItems = trainSchedules.map((schedule) => schedule.path).flat();
-  const pathItemQueries = [];
-  const pathItemSet = new Set<string>();
-  for (const item of pathItems) {
-    let query: SearchQuery;
-    if ('uic' in item) {
-      query = ['=', ['uic'], item.uic];
-      if (item.secondary_code) {
-        query = ['and', query, ['=', ['ch'], item.secondary_code]];
-      }
-    } else if ('trigram' in item) {
-      query = ['=', ['trigram'], item.trigram];
-      if (item.secondary_code) {
-        query = ['and', query, ['=', ['ch'], item.secondary_code]];
-      }
-    } else if ('operational_point' in item) {
-      query = ['=', ['obj_id'], item.operational_point];
-    } else {
-      continue; // track offset, handled by creating an empty node
-    }
-
-    // Avoid including the same query twice in the search payload
-    const key = JSON.stringify(query);
-    if (pathItemSet.has(key)) {
-      continue;
-    }
-
-    pathItemSet.add(key);
-    pathItemQueries.push(query);
-  }
-
-  if (pathItemQueries.length === 0) {
-    return null;
+type NodeIndex = {
+  node: MacroNodeResponse & { geocoord?: { lat: number; lng: number } };
+  saved: boolean;
+};
+export default class OsrdNgeSync {
+  /**
+   * Storing nodes by path item key
+   * It's the main storage for node.
+   * The saved attribut is to know if the data comes from the API
+   * If the value is a string, it's a key redirection
+   */
+  private nodesByPathKey: Record<string, NodeIndex | string>;
+
+  /**
+   * We keep a dictionnary of id/key to be able to find a node by its id
+   */
+  private nodesIdToKey: Record<number, string>;
+
+  /**
+   * Storing labels
+   */
+  private labels: Set<string>;
```

Comment:
This mixes up two kinds of labels: trainrun and node labels. In NGE, trainrun labels and node labels are in two separate buckets. The `labelRef` field indicates what kind of object the label can be attached to.

Note, up until now we were only handling trainrun labels, we were ignoring node labels.